### ISSUE TITLE
Resource Expiration/Rotation

### ISSUE THREAD

**Issue by vschettino-asimov** [pulumi/pulumi#13328]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Credential rotation is considered a cloud management best practice. The most common approach is to:
a) generate new credentials
b) update the dependent services accordingly
c) decommission the old credentials

This is exactly Pulumi's approach on Resource replacement operations, but AFAIK there's no way to force the replacement of a existing resource without changing one of this properties. Yes, I am aware of https://github.com/pulumi/pulumi/issues/6753.  `replaceOnChanges: ['*']` is a great feature but does not address "force a replacement" without a property change. Having to force a change only for that feels hacky and relies on humans to remember to take action. My specific use case is to rotate [DataDog API Keys](https://www.pulumi.com/registry/packages/datadog/api-docs/apikey/) but I'm sure it applies to many orders.

My first though about how to implement this is a `replaceIn` option, defined in `days` or other time lapse representations. Further deployments would always check this value + the last replacement and schedule a replacement if the criteria matches. An alternative would be providing a `pulumi replace --target URN` CLI, where dependent resources are automatically updated. In this approach users would be in charge of creating the rotation mechanism to call `pulumi replace` which is two folded: harder to use but more flexibility. Arguably, the `replaceIn` option is pretty simple and might not be suitable for all teams.

**Comment by justinvp** [comment]
Thanks for the suggestion, @vschettino-asimov!

**Comment by vschettino-asimov** [comment]
I just found out a `keepers` parameter on gcp.serviceAccount.Keys()` that does the job for that particular resource: https://www.pulumi.com/registry/packages/gcp/api-docs/serviceaccount/key/#keepers_nodejs

Could be useful for all types of resources

**Comment by Frassle** [comment]
So I think we have two other suggestions that would cover this, plus there's already a replace option for `up`.

One is general support for "keepers/triggers" which we have PR for https://github.com/pulumi/pulumi/pull/11089. This would let you write a trigger that would change it's value every X days for example, automatically replacing.

The other is the "tainting" resources https://github.com/pulumi/pulumi/issues/11657 where you could run something like `pulumi state taint <urn>` and the next update would replace that resource.

There's also the `--replace` argument to `up` that you could use today to force the engine to replace the resource.

**Comment by vschettino-asimov** [comment]
Thanks for the references @Frassle! Both `up --replace` and `state taint` are useful in this context and could be used with some additional tooling (like a scheduled CI job). I just want to emphasize that these approaches still demands an external mechanism to keep track of the rotation schedule, whereas in my suggestion it would be more "set and forget". With something like `replaceIn` in place regular deployments would take care of replacing these components.

**Comment by Frassle** [comment]
> With something like replaceIn in place regular deployments would take care of replacing these components.

I think triggers would cover that. You could write some code something like:
```
initialDate = Date(2023, 07, 12)
daysBetween = (Date.now - initialDate).TotalDays
trigger = daysBetween / 5 // trigger a replace every 5 days
res = new Resource("my-resource", args, ResourceOptions(Trigger=trigger))
```

That should be fairly expressible in any language and gives a lot of flexibility for when you'd want to fire the replace trigger.
It's also a lot more general than a dedicated resource option for "replace if last update was more than X time ago".

I suspect the only issue with the above is wanting to seed that logic based on old deployment information rather than fixing a start time in the program. I think a combination of state resources and lifecycle hooks _might_ cover that, but I'll have to think more about it.

In the mean time `pulumi up --replace` should work for you today.

**Comment by vschettino-asimov** [comment]
@Frassle yeah, I think your example would cover at least the most common use case. It does expose the `initialDate` instead  of letting Pulumi manage it internally, but probably it means more flexibility too.

For today I'll rely on `up --replace` but looking forward for having `trigger` as a more complete alternative.

**Comment by jkodroff** [comment]
Adding a few words of search spam so this issue is easier to find (b/c I missed it): Maximum age of a resource to trigger replacement
